xdsl/dialects/gpu.py

```diff
@@ -127,4 +194,64 @@ def get() -> ModuleEndOp:
         return ModuleEndOp.build()
 
 
-GPU = Dialect([ModuleOp, ModuleEndOp], [_GPUAttr])
+@irdl_op_definition
+class NumSubgroupsOp(Operation):
+    name = "gpu.num_subgroups"
+    result: Annotated[OpResult, IndexType]
+
+    @staticmethod
+    def get() -> NumSubgroupsOp:
+        return NumSubgroupsOp.build(result_types=[IndexType()])
+
+
+@irdl_op_definition
+class SetDefaultDeviceOp(Operation):
+    name = "gpu.set_default_device"
+    devIndex: Annotated[Operand, i32]
+
+    @staticmethod
+    def get(devIndex: SSAValue | Operation) -> SetDefaultDeviceOp:
+        return SetDefaultDeviceOp.build(operands=[SSAValue.get(devIndex)])
+
+
+@irdl_op_definition
+class SubgroupIdOp(Operation):
+    name = "gpu.subgroup_id"
+    result: Annotated[OpResult, IndexType]
+
+    @staticmethod
+    def get() -> SubgroupIdOp:
+        return SubgroupIdOp.build(result_types=[IndexType()])
+
+
+@irdl_op_definition
+class SubgroupSizeOp(Operation):
+    name = "gpu.subgroup_size"
+    result: Annotated[OpResult, IndexType]
+
+    @staticmethod
+    def get() -> SubgroupSizeOp:
+        return SubgroupSizeOp.build(result_types=[IndexType()])
+
+
+@irdl_op_definition
+class ThreadIdOp(Operation):
+    name = "gpu.thread_id"
+    dimension: OpAttr[DimensionAttr]
+    result: Annotated[OpResult, IndexType]
+
+    @staticmethod
+    def get(dim: DimensionAttr) -> ThreadIdOp:
+        return ThreadIdOp.build(result_types=[IndexType()],
+                                attributes={"dimension": dim})
+
+
+#_GPUAttr has to be registered instead of DimensionAttr and AllReduceOperationAttr here.
+# This is a hack to fit MLIR's syntax in xDSL's way of parsing attributes, without making GPU builtin.
+# Hopefully MLIR will parse it in a more xDSL-friendly way soon, so all that can be factored in proper xDSL
+# atrributes.
+GPU = Dialect([
+    BarrierOp, BlockDimOp, BlockIdOp, GlobalIdOp, GridDimOp, LaneIdOp,
+    ModuleOp, ModuleEndOp, NumSubgroupsOp, SetDefaultDeviceOp, SubgroupIdOp,
+    SubgroupSizeOp, ThreadIdOp
```

Comment:
Nit: Add commas at the end to make yapf have nice newlines

Reply:
If that is something we want to enforce, can we make yapf do it?
I hate having to manually check my formatter's formatting TBH.